### PR TITLE
[cli] better `ti test` to test single cpp file and no cpp test when testing python(s)

### PR DESCRIPTION
**Pull request by archibate**
Related issue = #684

usage example:
`ti test` to run all c++ test and python test.
`ti test --cpp` to run all c++ test.
`ti test --cpp spec` to run only the `spec` c++ test.
`ti test spec` to run only the `spec` python test.

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

### CODE REVIEW

**Comment by archibate** [comment]
Try `ti --cpp test alg_simp`? Seems it's a problem of `argparse` not setting up correctly and it's considering `alg_simp` as the same level as `test`...

**Comment by xumingkuan** [comment]
Still doesn't work.
```
C:\Users\xmk\Desktop\taichi\tests\cpp>ti --cpp test alg_simp
[Taichi] mode=development
[Taichi] version 0.5.10, cpu only, commit 5e896339, python 3.7.4

 *******************************************
 **     Taichi Programming Language       **
 *******************************************

Running C++ tests...
============================================================================
===
No tests ran


C:\Users\xmk\Desktop\taichi\tests\cpp>ti --cpp test test_alg_simp.cpp
[Taichi] mode=development
[Taichi] version 0.5.10, cpu only, commit 5e896339, python 3.7.4

 *******************************************
 **     Taichi Programming Language       **
 *******************************************

Running C++ tests...
============================================================================
===
No tests ran
```

**Comment by yuanming-hu** [comment]
This will be a very useful feature - We might need to take a look into https://github.com/taichi-dev/taichi/blob/d28533c503a1e4fd1101b549d63c6fb540618be2/taichi/core/testing.cpp#L11 to figure out how to test a specific C++ test.

File: docs/contributor_guide.rst [hunk 1]
```diff
@@ -136,7 +136,9 @@ Testing
 Tests should be added to ``taichi/tests``.
 Use ``ti test`` to run all the tests.
 Use ``ti test -v`` for verbose outputs.
-Use ``ti test <filename(s)>`` to run specific tests. Feel free to omit the ``test_`` prefix and ``.py`` suffix in the filenames.
+Use ``ti test <filename(s)>`` to run specific tests. e.g. ``ti test numpy_io`` and ``ti test test_numpy_io.py`` are equivalent.
+Use ``ti test -a <arch(s)>`` for test against specified architectures. e.g. ``ti test -a opengl`` or ``ti test numpy_io -a cuda,metal``.
+Use ``ti test -c`` to run only the C++ tests. e.g. ``ti test -c alg_simp``
```
Comment (archibate) [inline docs/contributor_guide.rst:141]: Was this the correct usage?

**Comment by yuanming-hu** [comment]
> How exactly test_cpp arguments work?

Ultimately your arguments are passed into https://github.com/catchorg/Catch2/blob/master/docs/command-line.md

**Comment by archibate** [comment]
What is a `Catch2`? One of our deps?

**Comment by archibate** [comment]
Could you just tell me what `<args>` look like in the legacy `ti test_cpp <args>`?

**Comment by yuanming-hu** [comment]
Maybe we should try this? https://github.com/catchorg/Catch2/blob/master/docs/command-line.md#naming-a-test-run

**Comment by yuanming-hu** [comment]
> What is a `Catch2`? One of our deps?

Yes.

**Comment by archibate** [comment]
> Maybe we should try this? https://github.com/catchorg/Catch2/blob/master/docs/command-line.md#naming-a-test-run

So I should call `test_cpp(['-n', 'alg_simp'])`?

**Comment by archibate** [comment]
Can't find `create_task` anywhere, wasn't `@unit('task')` call that?

File: python/taichi/main.py [hunk 2]
```diff
@@ -64,8 +61,10 @@ def test_cpp(args):
     # Cpp tests use the legacy non LLVM backend
     ti.reset()
     print("Running C++ tests...")
-    task = ti.Task('test')
+    task = ti.Task('test_gui')
     return int(task.run(*test_files))
```
Comment (archibate) [inline python/taichi/main.py:65]: I think I found the correct usage?

Comment (yuanming-hu) [reply python/taichi/main.py:65]: Sorry, but `test_gui` was for testing the GUI system.

**Comment by yuanming-hu** [comment]
> > Maybe we should try this? https://github.com/catchorg/Catch2/blob/master/docs/command-line.md#naming-a-test-run
> 
> So I should call `test_cpp(['-n', 'alg_simp'])`?

That sounds reasonable! We should try it.





> Can't find `create_task` anywhere, wasn't `@unit('task')` call that?

Please see https://github.com/taichi-dev/taichi/blob/d28533c503a1e4fd1101b549d63c6fb540618be2/taichi/system/run_tests.cpp#L18

File: docs/contributor_guide.rst (hunk 1 above)
Comment (yuanming-hu) [reply docs/contributor_guide.rst:141]: Since C++ tests run super fast so I haven't tried to run a single test only in the past...

**Comment by archibate** [comment]
So
https://github.com/taichi-dev/taichi/blob/d28533c503a1e4fd1101b549d63c6fb540618be2/taichi/core/testing.cpp#L19
applies `-n alg_simp`?

> That sounds reasonable! We should try it.

Tried, failed.

**Comment by archibate** [comment]
@xumingkuan I found the usage: `ti -c test simplify_add_zero`:
```
Running C++ tests...
!!!! ['simplify_add_zero']
===============================================================================
All tests passed (3 assertions in 1 test case)
```
This is because:
https://github.com/taichi-dev/taichi/blob/d28533c503a1e4fd1101b549d63c6fb540618be2/tests/cpp/test_alg_simp.cpp#L8

Seems catch2's granularity is function, while pytest is file.

**Comment by archibate** [comment]
Now I changed it to file granularity, please enjoy `ti -c test alg_simp`.

File: docs/contributor_guide.rst [hunk 3]
```diff
@@ -157,7 +157,9 @@ Testing
...
```
Comment (xumingkuan) [inline docs/contributor_guide.rst:162]: This should be `ti -c test alg_simp`.

Comment (archibate) [reply docs/contributor_guide.rst:162]: Fixed `argparse` happiness by poping `argv[1]`. Now it's `ti test -c alg_simp` legal now.

**Comment by archibate** [comment]
Should works now. Good night!

**Pull request merged**